src/Storages/MergeTree/MergeTreeSequentialSource.cpp

```diff
@@ -53,92 +56,65 @@ class MergeTreeSequentialSource : public ISource
 
     size_t getCurrentMark() const { return current_mark; }
 
-    size_t getCurrentRow() const { return current_row; }
-
 protected:
     Chunk generate() override;
 
 private:
+    void updateRowsToRead(size_t mark_number);
+
     const MergeTreeData & storage;
     StorageSnapshotPtr storage_snapshot;
+    MergeTreeReadTaskInfoPtr read_task_info;
 
-    /// Data part will not be removed if the pointer owns it
-    MergeTreeData::DataPartPtr data_part;
+    LoggerPtr log = getLogger("MergeTreeSequentialSource");
 
-    /// Alter and mutation commands that are required to be applied to the part on-fly.
-    AlterConversionsPtr alter_conversions;
+    MarkRanges mark_ranges;
+    std::shared_ptr<MarkCache> mark_cache;
 
-    /// Columns we have to read (each Block from read will contain them)
-    Names columns_to_read;
+    MergeTreeReadTask::Readers readers;
+    MergeTreeReadersChain readers_chain;
 
     /// Should read using direct IO
     bool read_with_direct_io;
 
-    LoggerPtr log = getLogger("MergeTreeSequentialSource");
-
-    std::optional<MarkRanges> mark_ranges;
-
-    std::shared_ptr<MarkCache> mark_cache;
-    using MergeTreeReaderPtr = std::unique_ptr<IMergeTreeReader>;
-    MergeTreeReaderPtr reader;
-
-    /// current mark at which we stop reading
+    /// Current mark at which we stop reading
     size_t current_mark = 0;
 
-    /// current row at which we stop reading
-    size_t current_row = 0;
+    /// Number of rows to read from current mark
+    size_t current_rows_to_read = 0;
 
     /// Closes readers and unlock part locks
     void finish();
 };
 
 MergeTreeSequentialSource::MergeTreeSequentialSource(
+    Block result_header,
     MergeTreeSequentialSourceType type,
     const MergeTreeData & storage_,
-    const StorageSnapshotPtr & storage_snapshot_,
-    MergeTreeData::DataPartPtr data_part_,
-    AlterConversionsPtr alter_conversions_,
-    Names columns_to_read_,
+    StorageSnapshotPtr storage_snapshot_,
+    MergeTreeReadTaskInfoPtr read_task_info_,
     std::optional<MarkRanges> mark_ranges_,
-    bool apply_deleted_mask,
     bool read_with_direct_io_,
     bool prefetch)
-    : ISource(storage_snapshot_->getSampleBlockForColumns(columns_to_read_))
+    : ISource(std::move(result_header))
     , storage(storage_)
-    , storage_snapshot(storage_snapshot_)
-    , data_part(std::move(data_part_))
-    , alter_conversions(std::move(alter_conversions_))
-    , columns_to_read(std::move(columns_to_read_))
-    , read_with_direct_io(read_with_direct_io_)
-    , mark_ranges(std::move(mark_ranges_))
+    , storage_snapshot(std::move(storage_snapshot_))
+    , read_task_info(std::move(read_task_info_))
+    , mark_ranges(mark_ranges_.value_or(MarkRanges{MarkRange(0, read_task_info->data_part->getMarksCount())}))
```

Comment:
```suggestion
    , mark_ranges(std::move(mark_ranges_.)value_or(MarkRanges{MarkRange(0, read_task_info->data_part->getMarksCount())}))
```

https://godbolt.org/z/4hcvqvsvz